pylint/checkers/modified_iterating_checker.py

```diff
@@ -0,0 +1,138 @@
+        ),
+    }
+
+    options = ()
+    priority = -2
+
+    @utils.check_messages(
+        "modified-iterating-list", "modified-iterating-dict", "modified-iterating-set"
+    )
+    def visit_for(self, node: nodes.For) -> None:
+        iter_obj = node.iter
+        for body_node in node.body:
+            if self._modified_iterating_list_cond(body_node, iter_obj):
+                self.add_message(
+                    "modified-iterating-list",
+                    node=body_node,
+                    args=(iter_obj.name,),
+                )
+            elif self._modified_iterating_dict_cond(body_node, iter_obj):
+                self.add_message(
+                    "modified-iterating-dict",
+                    node=body_node,
+                    args=(iter_obj.name,),
+                )
+            elif self._modified_iterating_set_cond(body_node, iter_obj):
+                self.add_message(
+                    "modified-iterating-set",
+                    node=body_node,
+                    args=(iter_obj.name,),
+                )
+
+    @staticmethod
+    def _is_attribute_call_expr(body_node) -> bool:
+        return (
+            isinstance(body_node, astroid.Expr)
+            and isinstance(body_node.value, astroid.Call)
+            and isinstance(body_node.value.func, astroid.Attribute)
+        )
+
+    @classmethod
+    def _common_cond_list_set(cls, node, list_obj, infer_val) -> bool:
+        return (infer_val == utils.safe_infer(list_obj)) and (
+            node.value.func.expr.name == list_obj.name
+        )
+
+    @staticmethod
+    def _dict_node_cond(node) -> bool:
+        return isinstance(node, astroid.Assign) and (
+            isinstance(node.targets[0], astroid.Subscript)
+        )
+
+    @classmethod
+    def _modified_iterating_list_cond(cls, node, list_obj) -> bool:
+        if not cls._is_attribute_call_expr(node):
+            return False
+        try:
+            infer_val = utils.safe_infer(node.value.func.expr)
+            return (
+                cls._common_cond_list_set(node, list_obj, infer_val)
+                and (infer_val is not None)
+                and (infer_val.pytype() == "builtins.list")
+                and node.value.func.attrname in {"append", "remove"}
+            )
+        except (astroid.InferenceError, AttributeError):
+            return False
+
+    @classmethod
+    def _modified_iterating_dict_cond(cls, node, list_obj) -> bool:
+        if not cls._dict_node_cond(node):
+            return False
+        try:
+            infer_val = utils.safe_infer(node.targets[0].value)
+            return (
+                (infer_val is not None)
+                and (infer_val.pytype() == "builtins.dict")
+                and (infer_val == list_obj.inferred()[0])
+                and (node.targets[0].value.name == list_obj.name)
+            )
+        except (astroid.InferenceError, AttributeError):
+            return False
```

Comment:
In this part of the code the except seems to be covered by the test... It makes me reconsider my other comment. Maybe the inference error come from ``list_obj.inferred()`` and the code can't raise an inference error for the 2 other functions ? If we can have an inference error for the other, we should add a test case where the inference fail and I think the try/except should be around the minimal amount of code (not safe_infer). If it can't raise an ``InferenceError`` we should remove the try.

```suggestion
        infer_val = utils.safe_infer(node.targets[0].value)
        try:
            return (
                (infer_val is not None)
                and (infer_val.pytype() == "builtins.dict")
                and (infer_val == list_obj.inferred()[0])
                and (node.targets[0].value.name == list_obj.name)
            )
        except (astroid.InferenceError, AttributeError):
            return False
```

Reply:
I think a cleaner solution would be to split the if statements. Do the first two checks in a separate if and then do a `safe_infer` on `list_obj` which can then lead to either `False` or `True`. We'll need to add 2 or 3 extra return statements but I think it reads much nicer and allows the removal of those `try...except`'s.

Reply:
Something like this ?
```suggestion
        infer_val = utils.safe_infer(node.targets[0].value)
        if infer_val is None or infer_val.pytype() != "builtins.dict":
            # Uninferable or not a dict
            return False
        if node.targets[0].value.name != list_obj.name:
            return False
        try:
            return infer_val == list_obj.inferred()[0]
        except (astroid.InferenceError, AttributeError):
            return False
```

Reply:
Yes, but then only with a `safe_infer` on L117 to remove the `try...except` (I think that should be possible).